composite_broker/src/main.rs

```diff
@@ -1,3 +1,706 @@
+            println!("\tConnecting...");
+            let ret_p = broker.accept_connect(addr.to_string(), p);  // returns ack
+            let mut ret_buf = [0u8; 128];      // will contain encoded bytes
+            ret_buf.write(&mut 0, ret_p).expect("Didn't write to buffer");// write to the buffer
+            socket.send_to(&ret_buf.as_mut(), addr).expect("Failed to send a response");
+        },
+        Message::Subscribe(p) => {
+            println!("\tSubscribing...");
+            let ret_p = broker.accept_sub(addr.to_string(), p);
+            let mut ret_buf = [0u8; 128];      // will contain encoded bytes
+            ret_buf.write(&mut 0, ret_p).expect("Didn't write to buffer"); // write to the buffer    
+            socket.send_to(&ret_buf.as_mut(), addr).expect("Failed to send to client");
+        },
+        Message::Publish(p) => {
+            println!("\tPublishing...");
+                let res = broker.accept_pub(addr.to_string(), p);
+                // encode the ack packet and the publish packet again
+                let client_list = &res.2;
+                let (mut ack_buf, mut pub_buf) = ([0u8; 128], [0u8; 128]);
+                ack_buf.write(&mut 0, res.0).expect("Didn't write to buffer"); 
+                pub_buf.write(&mut 0, res.1).expect("Didn't write to buffer");
+                // send the ack packet to this client
+                socket.send_to(&ack_buf.as_mut(), addr).expect("Failed to send to client");
+                for cli in client_list {
+                    println!("Sending to client...{}", cli);
+                    socket.send_to(&pub_buf.as_mut(), cli).expect("Failed to send to subscriber");
+                }
+        },
+        Message::Unsubscribe(p) => {
+            println!("\tUnsubscribing...");
+            let res = broker.accept_unsub(addr.to_string(), p);
+            let mut ret_buf = [0u8; 128];      // will contain encoded bytes
+            ret_buf.write(&mut 0, res).expect("Didn't write to buffer"); // write to the buffer    
+            socket.send_to(&ret_buf.as_mut(), addr).expect("Failed to send to client");
+            broker.get_sub_list();
+        }
+        _ => panic!("Incorrect type returned"),
+    };
+    // broker.clone()
 }
+
+fn main() -> io::Result<()>{
+
+    // make UDP socket
+    let socket = UdpSocket::bind("0.0.0.0:8888").expect("Could not bind socket");
+    // make the broker 
+    let broker: MBroker =MBroker::new();
+    let broker = Mutex::new(broker);
+    let broker = Arc::new(broker);
+
+    loop {
+        let mut buf = [0u8; 128];
+        
+        let sock = socket.try_clone().expect("Failed to clone socket");    // use socket clone to send to client
+        
+        match socket.recv_from(&mut buf) {  // receive the message into buffer
+            Ok((_, src)) => {
+                    println!("Handling incoming from {}", src);
+                    let thread_broker = broker.clone();
+                    thread::spawn(move || {
```

Comment:
As we've discussed, this is a thread per connection (or message). We really want the threads to be created at the start of the system, and have them all read from the socket. Only one of them should get each message. I think you're already working on this code, so I won't elaborate here.